AlphaWallet/Common/Views/ViewModels/AVPlayerViewModel.swift

```diff
@@ -0,0 +1,241 @@
+        case failure(Error)
+    }
+}
+
+extension AVAsset {
+
+    enum ImageGeneratorError: Error {
+        case `internal`(Error)
+    }
+
+    private func _generator(size: CGSize) -> AVAssetImageGenerator {
+        let generator = AVAssetImageGenerator(asset: self)
+        generator.maximumSize = size
+        generator.appliesPreferredTrackTransform = true
+        return generator
+    }
+
+    enum AVAssetError: Error {
+        case general(Error)
+        case unknown
+        case cancelled
+    }
+
+    enum AVAssetLoadingState: Equatable {
+        case loading
+        /// isVideoAvailable == false - suppose is audio, might be needed to update, good for now
+        case loaded(AVPlayerItem, isVideoAvailable: Bool)
+    }
+
+    func loadValuesAsync() -> AnyPublisher<AVAssetLoadingState, AVAssetError> {
+        let asset = self
+        let playableKey = "playable"
+        let durationKey = "duration"
+
+        return AnyPublisher<AVAssetLoadingState, AVAssetError>.create { seal in
+            seal.send(.loading)
+
+            asset.loadValuesAsynchronously(forKeys: [playableKey, durationKey]) {
+                DispatchQueue.main.async {
+                    var error: NSError?
+                    let status = asset.statusOfValue(forKey: playableKey, error: &error)
+
+                    switch status {
+                    case .loading:
+                        seal.send(.loading)
+                    case .loaded:
+                        let playerItem = AVPlayerItem(asset: asset)
+                        if let e = playerItem.error {
+                            seal.send(completion: .failure(AVAssetError.general(e)))
+                        } else {    
+                            seal.send(.loaded(playerItem, isVideoAvailable: asset.isVideoAvailable))
+                            seal.send(completion: .finished)
+                        }
+                    case .failed:
+                        let error = error.flatMap { AVAssetError.general($0) } ?? .unknown
+
+                        seal.send(completion: .failure(error))
+                    case .cancelled:
+                        seal.send(completion: .failure(.cancelled))
+                    case .unknown:
+                        seal.send(completion: .failure(.unknown))
+                    @unknown default:
+                        seal.send(completion: .failure(.unknown))
+                    }
+                }
+            }
+
+            return AnyCancellable {
+                asset.cancelLoading()
+            }
+        }.removeDuplicates()
+        .eraseToAnyPublisher()
+    }
+
+    func extractUIImageAsync(size: CGSize) -> AnyPublisher<UIImage?, ImageGeneratorError> {
+        let generator = _generator(size: size)
+        return AnyPublisher<UIImage?, ImageGeneratorError>.create { observer in
+            generator.generateCGImagesAsynchronously(forTimes: [NSValue(time: .zero)]) { (_, image, _, _, error) in
```

Comment:
In your tests, does `.zero` usually give good results?

Reply:
zero its a first frame of video

Reply:
Yes. But not all videos start immediately :)

Reply:
yes, we can tweak it later